presto-main/src/main/java/io/prestosql/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -228,11 +233,29 @@ public void enqueue(List<SerializedPage> pages)
         masterBuffer.addPages(serializedPageReferences);
 
         // process any pending reads from the client buffers
-        for (ClientBuffer clientBuffer : safeGetBuffersSnapshot()) {
+        clientBufferPollLoop();
+    }
+
+    /**
+     * Iterates client buffers, resuming where the previous cycle stopped if the master buffer was emptied before all clients were polled
+     * and no new client buffers have been inserted.
+     */
+    private void clientBufferPollLoop()
+    {
+        List<ClientBuffer> buffers = safeGetBuffersSnapshot();
+        if (buffers.isEmpty()) {
+            return;
+        }
+        // handle potential for racy update of indexOffset and buffer list copy
```

Comment:
`indexOffset` seems incorrect here (maybe was renamed after the comment was written). Maybe just say "next index" rather than referring to the field.

Reply:
You are correct, updated.